File: backend/data_store.py
import json
from pathlib import Path
from typing import Any

BASE_DIR = Path(__file__).parent
DATA_DIR = BASE_DIR / "data"
DATA_DIR.mkdir(parents=True, exist_ok=True)

FILES = {
    "notes": DATA_DIR / "notes.json",
    "grades": DATA_DIR / "grades.json",
    "assignments": DATA_DIR / "assignments.json",
    "announcements": DATA_DIR / "announcements.json",
}

for path in FILES.values():
    if not path.exists():
        path.write_text(json.dumps([], indent=2), encoding="utf-8")


def read_json(file_path: Path, default: Any = None) -> Any:
    if default is None:
        default = []
    if not file_path.exists():
        return default
    return json.loads(file_path.read_text(encoding="utf-8"))


def write_json(file_path: Path, data: Any) -> None:
    file_path.write_text(json.dumps(data, indent=2), encoding="utf-8")


def get_notes() -> list:
    return read_json(FILES["notes"])


def save_notes(notes: list) -> None:
    write_json(FILES["notes"], notes)


def get_grades() -> list:
    return read_json(FILES["grades"])


def save_grades(grades: list) -> None:
    write_json(FILES["grades"], grades)


def get_assignments() -> list:
    return read_json(FILES["assignments"])


def save_assignments(assignments: list) -> None:
    write_json(FILES["assignments"], assignments)


def get_announcements() -> list:
    return read_json(FILES["announcements"])


def save_announcements(announcements: list) -> None:
    write_json(FILES["announcements"], announcements)

File: backend/dataStore.js
const fs = require('fs').promises;
const path = require('path');

const dataDir = path.join(__dirname, 'data');
const files = {
  notes: path.join(dataDir, 'notes.json'),
  grades: path.join(dataDir, 'grades.json'),
  assignments: path.join(dataDir, 'assignments.json'),
  announcements: path.join(dataDir, 'announcements.json'),
  users: path.join(dataDir, 'users.json'),
  attendance: path.join(dataDir, 'attendance.json')
};

async function ensureDataFile(filePath, defaultValue) {
  try {
    await fs.access(filePath);
  } catch {
    await fs.writeFile(filePath, JSON.stringify(defaultValue, null, 2));
  }
}

async function init() {
  await Promise.all([
    ensureDataFile(files.notes, []),
    ensureDataFile(files.grades, []),
    ensureDataFile(files.assignments, []),
    ensureDataFile(files.announcements, []),
    ensureDataFile(files.users, []),
    ensureDataFile(files.attendance, [])
  ]);
}

async function readJson(filePath) {
  const raw = await fs.readFile(filePath, 'utf8');
  return JSON.parse(raw || '[]');
}

async function writeJson(filePath, data) {
  await fs.writeFile(filePath, JSON.stringify(data, null, 2));
}

module.exports = {
  init,
  async getNotes() {
    return readJson(files.notes);
  },
  async saveNotes(notes) {
    return writeJson(files.notes, notes);
  },
  async getGrades() {
    return readJson(files.grades);
  },
  async saveGrades(grades) {
    return writeJson(files.grades, grades);
  },
  async getAssignments() {
    return readJson(files.assignments);
  },
  async saveAssignments(assignments) {
    return writeJson(files.assignments, assignments);
  },
  async getAnnouncements() {
    return readJson(files.announcements);
  },
  async saveAnnouncements(announcements) {
    return writeJson(files.announcements, announcements);
  },
  async getUsers() {
    return readJson(files.users);
  },
  async saveUsers(users) {
    return writeJson(files.users, users);
  },
  async getAttendance() {
    return readJson(files.attendance);
  },
  async saveAttendance(attendance) {
    return writeJson(files.attendance, attendance);
  }
};
